### ISSUE TITLE
Receivers not getting messages from a full queue

### ISSUE THREAD

**Issue by myusuf3** [Azure/go-amqp#25]
Seems like we are running into a flow control issue with the library. We have been able to reason about what we think (we currently can't see a way this could happen) might solve the issue.  We have also been able to produce the problem with the attached sample code below, in under just couple hundred lines or so. 

We create a queue that we are simultaneously, producing messages and consuming message within the application. Through this process we are expecting the consumer to receive them as they are available on the queue to process and marked them as accepted. 

We setup a session and pass that to the `Producer` and `Consumer` which in turn  create their respective Sender and Receiver methods on that same session. 

On the `Consumer` we set LinkCredit to `27` more on that later. We are running `8` consumers in this example.  As you can see below. 

![image](https://user-images.githubusercontent.com/156825/108771933-53cdbb00-752a-11eb-924e-ae4d6cda6cf4.png)

What we are seeing it that the `Consumers` only manages to dequeue first 100 messages or so before the AMQP server stops providing any more messages for retrieval and we are stuck.  Bear in mind we are able to continue sending on that same session. 

We enabled debugging, and in the attached `logs.txt` you will see that  

`OutgoingWindow` reporting is kind of weird and sometimes returns max values in the logs. Is this a known issue? Can you shed some light on this? I spent quite some time trying to track the back and forth to illuminate what is happening on the wire. 

```
...
11:23:26.098572 TX: Begin{RemoteChannel: <nil>, NextOutgoingID: 0, IncomingWindow: 100, OutgoingWindow: 100, HandleMax: 4294967295, OfferedCapabilities: [], DesiredCapabilities: [], Properties: map[]}
11:23:26.098954 RX: Begin{RemoteChannel: 0, NextOutgoingID: 1, IncomingWindow: 16383, OutgoingWindow: 2147483647, HandleMax: 65535, OfferedCapabilities: [], DesiredCapabilities: [], Properties: map[]}
11:23:26.099082 TX: Attach{Name: nSbobxLLsp2GbySYOLWUFYDnf17WzICmn3gc3r5pmf9Pfuq46fbv7g, Handle: 0, Role: Sender, SenderSettleMode: <nil>, ReceiverSettleMode: <nil>, Source: <nil>, Target: source{Address: /queue-name, Durable: 0, ExpiryPolicy: , Timeout: 0, Dynamic: false, DynamicNodeProperties: map[], Capabilities: []}, Unsettled: map[], IncompleteUnsettled: false, InitialDeliveryCount: 0, MaxMessageSize: 0, OfferedCapabilities: [], DesiredCapabilities: [], Properties: map[]}
....
```

Now back to the numbers, this is a working theory we have that the `SessionIncomingWindow` which default `100` is used to control flow at the session level and the `LinkCredits` are being used at the link level. This is the part we are unsure about the relationship between the `SessionIncomingWindow` and `LinkCredits`. 

So in the example case here we have 8 consumers with `27` link credits each which is `216`. Which is little over the double of the `SessionIncomingWindow` which defaults to `100`  Since they are all sharing the same session, we assume this is what is tripping the control flow and causing the error we are seeing. 

If we lower just the link credits per consumer down to `24` which results to `192` we never see the lock up and all the messages a processed smoothly.  We were able to verify this with various combinations as well. 

![image](https://user-images.githubusercontent.com/156825/108773955-1159ad80-752d-11eb-842e-0e701a91edfe.png)

We are looking better to understand this relationship since it doesn't make sense currently. We are aware that these limits are being calculated on the fly based on other values until server and client can sync them. Could there be a race condition which ends up locking up the processing? 

I hope this is adequate information to help us debug the issue we are facing.  Let me know if you need anymore information. 

[amqp.zip](https://github.com/Azure/go-amqp/files/6025116/amqp.zip)

**Comment by serbrech** [comment]
Thank you for your detailed report. What version of go-amqp are you using?
We recently fixed a number of flow control issues (see latest PRs) that I believe might be the root cause of what you are seeing.

Could you retest using the latest go-amqp, and the HandleMessage func instead of the Receive?
I will do the test myself, but you might get to it before me :)

**Comment by myusuf3** [comment]
Have you had a chance to reproduce it? I tried it with the latest version available few days before I wrote up the issue and it continued to happen. But that might have been too old.  

It might be the disposition bug that could have been causing it. I will try with the latest and see.

**Comment by myusuf3** [comment]
So we had a similar issue again today where consumers where unable to get messages until we rebooted the consumer box. Would be really helpful to get some eyes on this issue. HandleMessage doesn't seem to resolve the issue either in the test project.

**Comment by myusuf3** [comment]
Any progress here?

**Comment by haraldkuegel** [comment]
Hello,

any new on this issue ?
i am facing the same problem. After #SessionIncomingWindow  messages are dequeued, the receiver doesn't fetch any more messages. Restarting the receiver works for another #SessionIncomingWindow messages.
The setup is 
- v0.13.11
- message handler api
- single sender / receiver
- Apache MQ (docker image with preconfigured settings)
- all messages are accepted (which Apache MQ shows in the web interface)
Since this is a very basic use case, i suspect that some settings (e.g. LinkOptions) might be incorrect.

Thanks!

**Comment by ogero** [comment]
Hi @jhendrixMSFT, it looks like https://github.com/Azure/go-amqp/pull/48 has resolved this ticket. 
At a glance, it seems the window underflow could explain what we are experiencing here, could you please confirm this?
Thanks!

**Comment by jhendrixMSFT** [comment]
Fixed in `v0.13.13`
